docs/pages/pmd/devdocs/pmd_devdocs_metrics_index.md

```diff
@@ -0,0 +1,107 @@
+
+*Operation metric, class metric.*
+
+### Description
+
+
+## Cyclomatic Complexity (CYCLO)
+
+*Operation metric.* Can be calculated on any non-abstract method.
+
+### Description
+
+Number of independent paths through a block of code \[[Lanza05](#Lanza05)\]. 
+Formally, given that the control flow graph of the block has n vertices, e edges and p connected components, the 
+Cyclomatic complexity of the block is given by `CYCLO = e - n + 2p` \[[McCabe76](#McCabe76)\]. In practice it can be 
+calculated by counting control flow statements following the standard rules given below.
+
+The standard version of the metric complies with McCabe's original definition:
+
+* Methods have a base complexity of 1.
+* +1 for every control flow statement (`if`, `case`, `catch`, `throw`, `do`, `while`, `for`, `break`, `continue`) and 
+conditional expression (`?:`) \[[Sonarqube](#Sonarqube)\]. Notice switch cases count as one, but not the switch itself: the point is that
+ a switch should have the same complexity value as the equivalent series of `if` statements.
+* `else`, `finally` and `default` don't count;
+* +1 for every boolean operator (`&&`, `||`) in the guard condition of a control flow statement. That's because
+Java has short-circuit evaluation semantics for boolean operators, which makes every boolean operator kind of a
+control flow statement in itself.
+ 
+### Versions
+
+* Version `CycloVersion#IGNORE_BOOLEAN_PATHS`: Boolean operators are not counted, nor are empty
+  fall-through cases in `switch` statements.
+ 
+ 
+## Lines of Code (LoC)
+
+*Operation metric, class metric.* Can be calculated on any of those nodes.
+
+### Description
+
+Simply counts the number of lines of code the operation or class takes up in the source. This metric doesn't discount
+ comments or blank lines.
+
+
+## Non-commenting source statements (NCSS)
+
+*Operation metric, class metric.* Can be calculated on any of those nodes.
+
+### Description
+
+Number of statements in a class or operation. That's roughly equivalent to counting the number of semicolons and 
+opening braces in the program. Comments and blank lines are ignored, and so are split statements (e.g. `int\n a;` counts
+ a single statement). 
+
+The standard version of the metric is based off JavaNCSS's version  \[[JavaNcss](#JavaNcss)\]:
+
+* +1 for any of the following statements: `if`, `else`, `while`, `do`, `for`, `switch`, `break`, `continue`, `return`, 
+`throw`, `synchronized`, `catch`, `finally`.
+* +1 for each assignment, variable declaration (except `for` loop initializers) or statement expression. We count 
+variables  declared on the same line (e.g. `int a, b, c;`) as a single statement. 
+* Contrary to Sonarqube, but as JavaNCSS, we count type declarations (class, interface, enum, annotation),  
+and method and field declarations \[[Sonarqube](#Sonarqube)\].
+* Contrary to JavaNCSS, but as Sonarqube, we do not count package declaration and import declarations as statements.
```

Comment:
Right, at least in theory. Using this definition, the Ncss count of a class is only the statements that are inside the declaration (plus the declaration statement), which I think makes extra sense since import statements are more metadata for the compiler than code. Plus, most IDEs fold them by default so they don't hamper readability nearly as much as eg many field declarations